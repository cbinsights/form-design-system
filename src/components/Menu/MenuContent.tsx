import React from 'react';
import cc from 'classcat';
import * as DropdownMenu from '@radix-ui/react-dropdown-menu';

export type MenuContentProps = Pick<
  DropdownMenu.DropdownMenuContentProps,
  | 'asChild'
  | 'children'
  | 'onCloseAutoFocus'
  | 'onEscapeKeyDown'
  | 'onFocusOutside'
  | 'onInteractOutside'
  | 'onPointerDownOutside'
  | 'portalled'
>;

const MenuContent = ({ children, ...rest }: MenuContentProps): JSX.Element => (
  <DropdownMenu.Content
    {...rest}
    className={cc(['fdsMenuItems elevation--2 rounded--all bgColor--white'])}
    align={'start'}
    avoidCollisions
    sideOffset={0}
    forceMount={undefined}
    loop={false}
    allowPinchZoom={false}
    side={'bottom'}
    collisionTolerance={0}
  >
    {children}
  </DropdownMenu.Content>
);

export default MenuContent;
