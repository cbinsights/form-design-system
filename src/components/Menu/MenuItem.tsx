import React from 'react';
import { noop } from 'util/index';
import * as DropdownMenu from '@radix-ui/react-dropdown-menu';
import cc from 'classcat';

export type MenuItemProps = Pick<
  DropdownMenu.DropdownMenuItemProps,
  'asChild' | 'disabled' | 'onSelect' | 'textValue' | 'children'
>;

const MenuItem = ({
  children,
  disabled,
  onSelect,
  ...rest
}: MenuItemProps): JSX.Element => (
  <DropdownMenu.Item
    className={cc(['fdsMenuItem', { 'fdsMenuItem--disabled': disabled }])}
    onSelect={disabled ? noop : onSelect}
    aria-disabled={disabled}
    disabled={disabled}
    {...rest}
  >
    {children}
  </DropdownMenu.Item>
);

export default MenuItem;
