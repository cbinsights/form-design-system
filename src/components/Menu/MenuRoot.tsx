import React from 'react';
import * as DropdownMenu from '@radix-ui/react-dropdown-menu';

export type MenuRootProps = Omit<DropdownMenu.DropdownMenuProps, 'modal' | 'dir'>;

const MenuRoot = ({ children, ...rest }: MenuRootProps): JSX.Element => {
  return (
    <DropdownMenu.Root {...rest} modal={true} dir={'ltr'}>
      {children}
    </DropdownMenu.Root>
  );
};

export default MenuRoot;
