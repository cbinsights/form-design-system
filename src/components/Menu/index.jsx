import React from 'react';
import PropTypes from 'prop-types';
import {
  Menu as ReachMenu,
  MenuItems as ReachMenuItems,
  MenuButton as ReachMenuButton,
  MenuPopover as ReachMenuPopover,
} from '@reach/menu-button';

const Menu = ({ isDisabled = false, children, trigger }) => {
  if (isDisabled) return trigger;

  return (
    <ReachMenu>
      <ReachMenuButton as="div">{trigger}</ReachMenuButton>
      <ReachMenuPopover className="elevation--2 rounded--all bgColor--white">
        <ReachMenuItems>{children}</ReachMenuItems>
      </ReachMenuPopover>
    </ReachMenu>
  );
};

Menu.propTypes = {
  /** JSX node - the element that should trigger the menu */
  trigger: PropTypes.node.isRequired,

  /** React children (should be of type `MenuItem`) */
  children: PropTypes.oneOfType([PropTypes.arrayOf(PropTypes.node), PropTypes.node])
    .isRequired,

  /** Prevent trigger from opening menu */
  isDisabled: PropTypes.bool,
};

export default Menu;
