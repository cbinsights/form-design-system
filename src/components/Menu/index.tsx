import React from 'react';
import Content from './MenuContent';
import Item from './MenuItem';
import Trigger from './MenuTrigger';
import * as DropdownMenu from '@radix-ui/react-dropdown-menu';

export type MenuProps = Omit<DropdownMenu.DropdownMenuProps, 'modal' | 'dir'>;

const Menu = ({ children, ...rest }: MenuProps): JSX.Element => {
  return (
    <DropdownMenu.Root {...rest} modal={true} dir={'ltr'}>
      {children}
    </DropdownMenu.Root>
  );
};

Menu.Content = Content;
Menu.Item = Item;
Menu.Trigger = Trigger;

export default Menu;
