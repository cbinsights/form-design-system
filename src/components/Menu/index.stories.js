import React from 'react';
import { action } from '@storybook/addon-actions';

import Button from 'components/Button';

import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';

import CloneIcon from 'icons/react/CloneIcon';
import TrashIcon from 'icons/react/TrashIcon';
import EditIcon from 'icons/react/EditIcon';
import FundingIcon from 'icons/react/FundingIcon';
import Tooltip from 'components/Tooltip';

import { MenuItem, MenuLink } from 'components';
import Menu from '.';

export const Primary = (args) => (
  <Menu {...args} trigger={<Button label="Menu trigger" hasCaret />}>
    <MenuItem onSelect={action('copy')}>Copy</MenuItem>
    <MenuItem onSelect={action('delete')}>Delete</MenuItem>
  </Menu>
);

export const TriggerElement = (args) => (
  <Flex>
    <FlexItem>
      <Menu
        {...args}
        trigger={<Button theme="outlined" label="Recommended trigger" hasCaret />}
      >
        <MenuItem onSelect={action('🍰')}>Cake</MenuItem>
        <MenuItem onSelect={action('🍕')}>Pizza</MenuItem>
        <MenuItem onSelect={action('☠️')}>Tide pod</MenuItem>
      </Menu>
    </FlexItem>
    <FlexItem>
      <Menu {...args} trigger={<button>any button will work</button>}>
        <MenuItem onSelect={action('🍰')}>Cake</MenuItem>
        <MenuItem onSelect={action('🍕')}>Pizza</MenuItem>
        <MenuItem onSelect={action('☠️')}>Tide pod</MenuItem>
      </Menu>
    </FlexItem>
  </Flex>
);

TriggerElement.parameters = {
  docs: {
    description: {
      story:
        'A `Button` with the `hasCaret` prop enabled is recommended, but the Menu trigger can accept a `button`, or any component that renders a focusable element.',
    },
  },
};

export const CustomItems = (args) => (
  <Menu {...args} trigger={<Button hasCaret label="actions" theme="outlined" />}>
    <MenuItem onSelect={action('Delete')}>
      <div className="color--primary">
        <Flex align="center">
          <FlexItem shrink>
            <TrashIcon />
          </FlexItem>
          <FlexItem>Delete</FlexItem>
        </Flex>
      </div>
    </MenuItem>
    <MenuItem onSelect={action('Edit')}>
      <div className="color--primary">
        <Flex align="center">
          <FlexItem shrink>
            <EditIcon />
          </FlexItem>
          <FlexItem>Edit</FlexItem>
        </Flex>
      </div>
    </MenuItem>
    <MenuItem onSelect={action('Clone')}>
      <div className="color--primary">
        <Flex align="center">
          <FlexItem shrink>
            <CloneIcon />
          </FlexItem>
          <FlexItem>Clone</FlexItem>
        </Flex>
      </div>
    </MenuItem>
    <MenuItem onSelect={action('Bribe')}>
      <div className="color--primary">
        <Flex align="center">
          <FlexItem shrink>
            <FundingIcon />
          </FlexItem>
          <FlexItem>Bribe</FlexItem>
        </Flex>
      </div>
    </MenuItem>
    <MenuLink onSelect={action('Storybook')} href="/">
      Go to Storybook home
    </MenuLink>
  </Menu>
);

CustomItems.parameters = {
  docs: {
    description: {
      story: '`MenuItem` and `MenuLink` will accept a plain string or any JSX node.',
    },
  },
};

export const Rest = (args) => (
  <Menu {...args} trigger={<Button hasCaret label="actions" theme="outlined" />}>
    <MenuItem data-test="hello">Delete</MenuItem>
    <MenuLink href="#" data-test="world">
      Edit
    </MenuLink>
  </Menu>
);

Rest.parameters = {
  docs: {
    description: {
      story: '`MenuItem` and `MenuLink` accept rest props.',
    },
  },
};

export const DisabledItems = (args) => (
  <Menu
    {...args}
    trigger={<Button theme="outlined" label="Menu with disabled item" hasCaret />}
  >
    <MenuItem
      isDisabled
      onSelect={() => {
        action('🍰');
      }}
    >
      <Tooltip trigger={<button>Cake</button>} message="Not your birthday" />
    </MenuItem>
    <MenuItem
      onSelect={() => {
        action('🍕');
      }}
    >
      Pizza
    </MenuItem>
    <MenuItem
      onSelect={() => {
        action('☠️');
      }}
    >
      Tide pod
    </MenuItem>
  </Menu>
);

DisabledItems.parameters = {
  docs: {
    description: {
      story:
        'When the `isDisabled` prop is passed to a `MenuItem`, the item will be styled as disabled and the `onSelect` handler will not be called. Prefer hiding actions instead of disabling them. Only disable a `MenuItem` if the disabling is only temporary (e.g. disabling a collections clone while a clone is in progress). When disabling a `MenuItem`, a `Tooltip` is recommended to explain why the action is disabled.',
    },
  },
};

export default {
  component: Menu,
  subcomponents: { MenuItem, MenuLink },
  title: 'components/Menu',
  parameters: {
    componentSubtitle:
      'Renders a dropdown containing links, and actions the user can take.',
    docs: {
      description: {
        component:
          'Use FDS `Popover` instead for content like forms & arbitrary content. Use `MenuLink` for rendering links, and default to using `MenuItem` otherwise.',
      },
    },
  },
};
