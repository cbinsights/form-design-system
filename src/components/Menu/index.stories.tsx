import React from 'react';
import Button from 'components/Button';
import Flex from 'components/Flex';
import FlexItem from 'components/FlexItem';
import CloneIcon from 'icons/react/CloneIcon';
import TrashIcon from 'icons/react/TrashIcon';
import EditIcon from 'icons/react/EditIcon';
import FundingIcon from 'icons/react/FundingIcon';
import Tooltip from 'components/Tooltip';
import { MenuItem, MenuLink } from 'components';
import Menu, { MenuProps } from '.';
import { StoryObj } from '@storybook/react';

interface MenuStoryProps extends MenuProps {
  onSelect: () => void;
}

export const Primary: StoryObj<MenuStoryProps> = {
  render: ({ onSelect, ...args }: MenuStoryProps): JSX.Element => (
    <Menu {...args} trigger={<Button label="Menu trigger" hasCaret />}>
      <MenuItem onSelect={onSelect}>Copy</MenuItem>
      <MenuItem onSelect={onSelect}>Delete</MenuItem>
    </Menu>
  ),
};

export const TriggerElement: StoryObj<MenuStoryProps> = {
  render: ({ onSelect, ...args }: MenuStoryProps): JSX.Element => (
    <Flex>
      <FlexItem>
        <Menu
          {...args}
          trigger={<Button theme="outlined" label="Recommended trigger" hasCaret />}
        >
          <MenuItem onSelect={onSelect}>Cake</MenuItem>
          <MenuItem onSelect={onSelect}>Pizza</MenuItem>
          <MenuItem onSelect={onSelect}>Tide pod</MenuItem>
        </Menu>
      </FlexItem>
      <FlexItem>
        <Menu {...args} trigger={<button>any button will work</button>}>
          <MenuItem onSelect={onSelect}>Cake</MenuItem>
          <MenuItem onSelect={onSelect}>Pizza</MenuItem>
          <MenuItem onSelect={onSelect}>Tide pod</MenuItem>
        </Menu>
      </FlexItem>
    </Flex>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'A `Button` with the `hasCaret` prop enabled is recommended, but the Menu trigger can accept a `button`, or any component that renders a focusable element.',
      },
    },
  },
};

export const CustomItems: StoryObj<MenuStoryProps> = {
  render: ({ onSelect, ...args }: MenuStoryProps): JSX.Element => (
    <Menu {...args} trigger={<Button hasCaret label="actions" theme="outlined" />}>
      <MenuItem onSelect={onSelect}>
        <div className="color--primary">
          <Flex align="center">
            <FlexItem shrink>
              <TrashIcon />
            </FlexItem>
            <FlexItem>Delete</FlexItem>
          </Flex>
        </div>
      </MenuItem>
      <MenuItem onSelect={onSelect}>
        <div className="color--primary">
          <Flex align="center">
            <FlexItem shrink>
              <EditIcon />
            </FlexItem>
            <FlexItem>Edit</FlexItem>
          </Flex>
        </div>
      </MenuItem>
      <MenuItem onSelect={onSelect}>
        <div className="color--primary">
          <Flex align="center">
            <FlexItem shrink>
              <CloneIcon />
            </FlexItem>
            <FlexItem>Clone</FlexItem>
          </Flex>
        </div>
      </MenuItem>
      <MenuItem onSelect={onSelect}>
        <div className="color--primary">
          <Flex align="center">
            <FlexItem shrink>
              <FundingIcon />
            </FlexItem>
            <FlexItem>Bribe</FlexItem>
          </Flex>
        </div>
      </MenuItem>
      <MenuLink onSelect={onSelect} href="/">
        Go to Storybook home
      </MenuLink>
    </Menu>
  ),
  parameters: {
    docs: {
      description: {
        story: '`MenuItem` and `MenuLink` will accept a plain string or any JSX node.',
      },
    },
  },
};

export const Rest: StoryObj<MenuStoryProps> = {
  render: (args: MenuProps): JSX.Element => (
    <Menu {...args} trigger={<Button hasCaret label="actions" theme="outlined" />}>
      <MenuItem data-test="hello" onSelect={() => {}}>
        Delete
      </MenuItem>
      <MenuLink href="#" data-test="world">
        Edit
      </MenuLink>
    </Menu>
  ),
  parameters: {
    docs: {
      description: {
        story: '`MenuItem` and `MenuLink` accept rest props.',
      },
    },
  },
};

export const DisabledItems: StoryObj<MenuStoryProps> = {
  render: ({ onSelect, ...args }: MenuStoryProps): JSX.Element => (
    <Menu
      {...args}
      trigger={<Button theme="outlined" label="Menu with disabled item" hasCaret />}
    >
      <MenuItem isDisabled onSelect={onSelect}>
        <Tooltip trigger={<button>Cake</button>} message="Not your birthday" />
      </MenuItem>
      <MenuItem onSelect={onSelect}>Pizza</MenuItem>
      <MenuItem onSelect={onSelect}>Tide pod</MenuItem>
    </Menu>
  ),
  parameters: {
    docs: {
      description: {
        story:
          'When the `isDisabled` prop is passed to a `MenuItem`, the item will be styled as disabled and the `onSelect` handler will not be called. Prefer hiding actions instead of disabling them. Only disable a `MenuItem` if the disabling is only temporary (e.g. disabling a collections clone while a clone is in progress). When disabling a `MenuItem`, a `Tooltip` is recommended to explain why the action is disabled.',
      },
    },
  },
};

export default {
  component: Menu,
  subcomponents: {
    MenuItem,
    MenuLink,
  },
  title: 'Components/Menu',
  argTypes: {
    onSelect: {
      action: 'onSelect',
    },
  },
  parameters: {
    componentSubtitle:
      'Renders a dropdown containing links, and actions the user can take.',
    docs: {
      description: {
        component:
          'Use FDS `Popover` instead for content like forms & arbitrary content. Use `MenuLink` for rendering links, and default to using `MenuItem` otherwise.',
      },
    },
  },
};
