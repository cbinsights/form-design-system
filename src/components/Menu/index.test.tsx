import React from 'react';
import { render, screen } from '@testing-library/react';

import { MenuItem } from 'components';
import Menu from '.';

describe('Menu component', () => {
  /**
   * `@reach/menu-button` has its own test coverage.
   * FDS provides only styling overrides, so snapshots will be adequate coverage for our code.
   */
  it('renders component (strings)', () => {
    render(
      <Menu trigger={<a href="#">trigger</a>}>
        <MenuItem onSelect={jest.fn()}>First</MenuItem>
        <MenuItem onSelect={jest.fn()}>Second</MenuItem>
        <MenuItem onSelect={jest.fn()}>Third</MenuItem>
      </Menu>
    );
    expect(screen.getByText('First')).toBeInTheDocument();
    expect(screen.getByText('Second')).toBeInTheDocument();
    expect(screen.getByText('Third')).toBeInTheDocument();
  });
  it('renders component (JSX items)', () => {
    render(
      <Menu trigger={<a href="#">trigger</a>}>
        <MenuItem onSelect={jest.fn()}>
          <p>First</p>
        </MenuItem>
        <MenuItem onSelect={jest.fn()}>
          <p>Second</p>
        </MenuItem>
        <MenuItem onSelect={jest.fn()}>
          <p>Third</p>
        </MenuItem>
      </Menu>
    );
    expect(screen.getByText('First')).toBeInTheDocument();
    expect(screen.getByText('Second')).toBeInTheDocument();
    expect(screen.getByText('Third')).toBeInTheDocument();
  });
  it('renders trigger element only', () => {
    const mockOnClick = jest.fn();
    render(
      <Menu
        isDisabled
        trigger={
          <a href="#" onClick={mockOnClick}>
            trigger
          </a>
        }
      >
        <MenuItem onSelect={jest.fn()}>First</MenuItem>
        <MenuItem onSelect={jest.fn()}>Second</MenuItem>
        <MenuItem onSelect={jest.fn()}>Third</MenuItem>
      </Menu>
    );
    const anchor = screen.getByRole('link');
    expect(anchor).toBeInTheDocument();
    // MenuItems are not expected to be rendered so we check for them to be null
    expect(screen.queryByText('First')).toBeNull();
    expect(screen.queryByText('Second')).toBeNull();
    expect(screen.queryByText('Third')).toBeNull();
  });
});
