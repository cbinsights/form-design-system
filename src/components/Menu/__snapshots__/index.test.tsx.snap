// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Popover component matches snapshot (JSX items) 1`] = `
<Menu
  trigger={
    <a
      href="#"
    >
      trigger
    </a>
  }
>
  <Menu>
    <DropdownProvider>
      <DescendantProvider
        context={
          Object {
            "$$typeof": Symbol(react.context),
            "Consumer": Object {
              "$$typeof": Symbol(react.context),
              "_calculateChangedBits": null,
              "_context": [Circular],
            },
            "Provider": Object {
              "$$typeof": Symbol(react.provider),
              "_context": [Circular],
            },
            "_calculateChangedBits": null,
            "_currentRenderer": Object {},
            "_currentRenderer2": null,
            "_currentValue": Object {
              "descendants": Array [],
              "registerDescendant": [Function],
              "unregisterDescendant": [Function],
            },
            "_currentValue2": Object {
              "descendants": Array [],
              "registerDescendant": [Function],
              "unregisterDescendant": [Function],
            },
            "_threadCount": 0,
            "displayName": "DropdownDescendantContext",
          }
        }
        items={
          Array [
            Object {
              "disabled": undefined,
              "element": <div
                data-reach-menu-item=""
                data-valuetext="First"
                id="option-0--menu--5"
                role="menuitem"
                tabindex="-1"
              >
                <p>
                  First
                </p>
              </div>,
              "index": 0,
              "isLink": false,
              "key": "First",
            },
            Object {
              "disabled": undefined,
              "element": <div
                data-reach-menu-item=""
                data-valuetext="Second"
                id="option-1--menu--5"
                role="menuitem"
                tabindex="-1"
              >
                <p>
                  Second
                </p>
              </div>,
              "index": 1,
              "isLink": false,
              "key": "Second",
            },
            Object {
              "disabled": undefined,
              "element": <div
                data-reach-menu-item=""
                data-valuetext="Third"
                id="option-2--menu--5"
                role="menuitem"
                tabindex="-1"
              >
                <p>
                  Third
                </p>
              </div>,
              "index": 2,
              "isLink": false,
              "key": "Third",
            },
          ]
        }
        set={[Function]}
      >
        <MenuButton
          as="a"
          href="#"
        >
          <a
            aria-controls="menu--5"
            aria-haspopup={true}
            data-reach-menu-button=""
            href="#"
            id="menu-button--menu--5"
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            type="button"
          >
            trigger
          </a>
        </MenuButton>
        <MenuPopover
          className="elevation--2 rounded--all bgColor--white"
        >
          <Popover
            as="div"
            className="elevation--2 rounded--all bgColor--white"
            data-reach-menu-popover=""
            hidden={true}
            onBlur={[Function]}
            targetRef={
              Object {
                "current": <a
                  aria-controls="menu--5"
                  aria-haspopup="true"
                  data-reach-menu-button=""
                  href="#"
                  id="menu-button--menu--5"
                  type="button"
                >
                  trigger
                </a>,
              }
            }
          >
            <Portal>
              <Portal
                containerInfo={
                  <reach-portal>
                    <div
                      class="elevation--2 rounded--all bgColor--white"
                      data-reach-menu-popover=""
                      data-reach-popover=""
                      hidden=""
                      style="position: absolute; left: 0px; top: 0px;"
                    >
                      <div
                        aria-labelledby="menu-button--menu--5"
                        data-reach-menu-items=""
                        id="menu--5"
                        role="menu"
                        tabindex="-1"
                      >
                        <div
                          data-reach-menu-item=""
                          data-valuetext="First"
                          id="option-0--menu--5"
                          role="menuitem"
                          tabindex="-1"
                        >
                          <p>
                            First
                          </p>
                        </div>
                        <div
                          data-reach-menu-item=""
                          data-valuetext="Second"
                          id="option-1--menu--5"
                          role="menuitem"
                          tabindex="-1"
                        >
                          <p>
                            Second
                          </p>
                        </div>
                        <div
                          data-reach-menu-item=""
                          data-valuetext="Third"
                          id="option-2--menu--5"
                          role="menuitem"
                          tabindex="-1"
                        >
                          <p>
                            Third
                          </p>
                        </div>
                      </div>
                    </div>
                  </reach-portal>
                }
              >
                <PopoverImpl
                  as="div"
                  className="elevation--2 rounded--all bgColor--white"
                  data-reach-menu-popover=""
                  hidden={true}
                  onBlur={[Function]}
                  targetRef={
                    Object {
                      "current": <a
                        aria-controls="menu--5"
                        aria-haspopup="true"
                        data-reach-menu-button=""
                        href="#"
                        id="menu-button--menu--5"
                        type="button"
                      >
                        trigger
                      </a>,
                    }
                  }
                >
                  <div
                    className="elevation--2 rounded--all bgColor--white"
                    data-reach-menu-popover=""
                    data-reach-popover=""
                    hidden={true}
                    onBlur={[Function]}
                    style={
                      Object {
                        "left": "0px",
                        "position": "absolute",
                        "top": "0px",
                      }
                    }
                  >
                    <MenuItems>
                      <div
                        aria-labelledby="menu-button--menu--5"
                        data-reach-menu-items=""
                        id="menu--5"
                        onKeyDown={[Function]}
                        role="menu"
                        tabIndex={-1}
                      >
                        <MenuItem
                          onSelect={[Function]}
                        >
                          <MenuItem
                            onSelect={[Function]}
                          >
                            <ForwardRef
                              as="div"
                              onSelect={[Function]}
                            >
                              <div
                                data-reach-menu-item=""
                                data-valuetext="First"
                                id="option-0--menu--5"
                                onClick={[Function]}
                                onDragStart={[Function]}
                                onFocus={[Function]}
                                onMouseDown={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                onMouseMove={[Function]}
                                onMouseUp={[Function]}
                                role="menuitem"
                                tabIndex={-1}
                              >
                                <p>
                                  First
                                </p>
                              </div>
                            </ForwardRef>
                          </MenuItem>
                        </MenuItem>
                        <MenuItem
                          onSelect={[Function]}
                        >
                          <MenuItem
                            onSelect={[Function]}
                          >
                            <ForwardRef
                              as="div"
                              onSelect={[Function]}
                            >
                              <div
                                data-reach-menu-item=""
                                data-valuetext="Second"
                                id="option-1--menu--5"
                                onClick={[Function]}
                                onDragStart={[Function]}
                                onFocus={[Function]}
                                onMouseDown={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                onMouseMove={[Function]}
                                onMouseUp={[Function]}
                                role="menuitem"
                                tabIndex={-1}
                              >
                                <p>
                                  Second
                                </p>
                              </div>
                            </ForwardRef>
                          </MenuItem>
                        </MenuItem>
                        <MenuItem
                          onSelect={[Function]}
                        >
                          <MenuItem
                            onSelect={[Function]}
                          >
                            <ForwardRef
                              as="div"
                              onSelect={[Function]}
                            >
                              <div
                                data-reach-menu-item=""
                                data-valuetext="Third"
                                id="option-2--menu--5"
                                onClick={[Function]}
                                onDragStart={[Function]}
                                onFocus={[Function]}
                                onMouseDown={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                onMouseMove={[Function]}
                                onMouseUp={[Function]}
                                role="menuitem"
                                tabIndex={-1}
                              >
                                <p>
                                  Third
                                </p>
                              </div>
                            </ForwardRef>
                          </MenuItem>
                        </MenuItem>
                      </div>
                    </MenuItems>
                  </div>
                </PopoverImpl>
              </Portal>
            </Portal>
          </Popover>
        </MenuPopover>
      </DescendantProvider>
    </DropdownProvider>
  </Menu>
</Menu>
`;

exports[`Popover component matches snapshot (disabled) 1`] = `
<Menu
  isDisabled={true}
  trigger={
    <a
      href="#"
    >
      trigger
    </a>
  }
>
  <a
    href="#"
  >
    trigger
  </a>
</Menu>
`;

exports[`Popover component matches snapshot (string items) 1`] = `
<Menu
  trigger={
    <a
      href="#"
    >
      trigger
    </a>
  }
>
  <Menu>
    <DropdownProvider>
      <DescendantProvider
        context={
          Object {
            "$$typeof": Symbol(react.context),
            "Consumer": Object {
              "$$typeof": Symbol(react.context),
              "_calculateChangedBits": null,
              "_context": [Circular],
            },
            "Provider": Object {
              "$$typeof": Symbol(react.provider),
              "_context": [Circular],
            },
            "_calculateChangedBits": null,
            "_currentRenderer": Object {},
            "_currentRenderer2": null,
            "_currentValue": Object {
              "descendants": Array [],
              "registerDescendant": [Function],
              "unregisterDescendant": [Function],
            },
            "_currentValue2": Object {
              "descendants": Array [],
              "registerDescendant": [Function],
              "unregisterDescendant": [Function],
            },
            "_threadCount": 0,
            "displayName": "DropdownDescendantContext",
          }
        }
        items={
          Array [
            Object {
              "disabled": undefined,
              "element": <div
                data-reach-menu-item=""
                data-valuetext="First"
                id="option-0--menu--1"
                role="menuitem"
                tabindex="-1"
              >
                First
              </div>,
              "index": 0,
              "isLink": false,
              "key": "First",
            },
            Object {
              "disabled": undefined,
              "element": <div
                data-reach-menu-item=""
                data-valuetext="Second"
                id="option-1--menu--1"
                role="menuitem"
                tabindex="-1"
              >
                Second
              </div>,
              "index": 1,
              "isLink": false,
              "key": "Second",
            },
            Object {
              "disabled": undefined,
              "element": <div
                data-reach-menu-item=""
                data-valuetext="Third"
                id="option-2--menu--1"
                role="menuitem"
                tabindex="-1"
              >
                Third
              </div>,
              "index": 2,
              "isLink": false,
              "key": "Third",
            },
          ]
        }
        set={[Function]}
      >
        <MenuButton
          as="a"
          href="#"
        >
          <a
            aria-controls="menu--1"
            aria-haspopup={true}
            data-reach-menu-button=""
            href="#"
            id="menu-button--menu"
            onKeyDown={[Function]}
            onMouseDown={[Function]}
            type="button"
          >
            trigger
          </a>
        </MenuButton>
        <MenuPopover
          className="elevation--2 rounded--all bgColor--white"
        >
          <Popover
            as="div"
            className="elevation--2 rounded--all bgColor--white"
            data-reach-menu-popover=""
            hidden={true}
            onBlur={[Function]}
            targetRef={
              Object {
                "current": <a
                  aria-controls="menu--1"
                  aria-haspopup="true"
                  data-reach-menu-button=""
                  href="#"
                  id="menu-button--menu"
                  type="button"
                >
                  trigger
                </a>,
              }
            }
          >
            <Portal>
              <Portal
                containerInfo={
                  <reach-portal>
                    <div
                      class="elevation--2 rounded--all bgColor--white"
                      data-reach-menu-popover=""
                      data-reach-popover=""
                      hidden=""
                      style="position: absolute; left: 0px; top: 0px;"
                    >
                      <div
                        aria-labelledby="menu-button--menu"
                        data-reach-menu-items=""
                        id="menu--1"
                        role="menu"
                        tabindex="-1"
                      >
                        <div
                          data-reach-menu-item=""
                          data-valuetext="First"
                          id="option-0--menu--1"
                          role="menuitem"
                          tabindex="-1"
                        >
                          First
                        </div>
                        <div
                          data-reach-menu-item=""
                          data-valuetext="Second"
                          id="option-1--menu--1"
                          role="menuitem"
                          tabindex="-1"
                        >
                          Second
                        </div>
                        <div
                          data-reach-menu-item=""
                          data-valuetext="Third"
                          id="option-2--menu--1"
                          role="menuitem"
                          tabindex="-1"
                        >
                          Third
                        </div>
                      </div>
                    </div>
                  </reach-portal>
                }
              >
                <PopoverImpl
                  as="div"
                  className="elevation--2 rounded--all bgColor--white"
                  data-reach-menu-popover=""
                  hidden={true}
                  onBlur={[Function]}
                  targetRef={
                    Object {
                      "current": <a
                        aria-controls="menu--1"
                        aria-haspopup="true"
                        data-reach-menu-button=""
                        href="#"
                        id="menu-button--menu"
                        type="button"
                      >
                        trigger
                      </a>,
                    }
                  }
                >
                  <div
                    className="elevation--2 rounded--all bgColor--white"
                    data-reach-menu-popover=""
                    data-reach-popover=""
                    hidden={true}
                    onBlur={[Function]}
                    style={
                      Object {
                        "left": "0px",
                        "position": "absolute",
                        "top": "0px",
                      }
                    }
                  >
                    <MenuItems>
                      <div
                        aria-labelledby="menu-button--menu"
                        data-reach-menu-items=""
                        id="menu--1"
                        onKeyDown={[Function]}
                        role="menu"
                        tabIndex={-1}
                      >
                        <MenuItem
                          onSelect={[Function]}
                        >
                          <MenuItem
                            onSelect={[Function]}
                          >
                            <ForwardRef
                              as="div"
                              onSelect={[Function]}
                            >
                              <div
                                data-reach-menu-item=""
                                data-valuetext="First"
                                id="option-0--menu--1"
                                onClick={[Function]}
                                onDragStart={[Function]}
                                onFocus={[Function]}
                                onMouseDown={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                onMouseMove={[Function]}
                                onMouseUp={[Function]}
                                role="menuitem"
                                tabIndex={-1}
                              >
                                First
                              </div>
                            </ForwardRef>
                          </MenuItem>
                        </MenuItem>
                        <MenuItem
                          onSelect={[Function]}
                        >
                          <MenuItem
                            onSelect={[Function]}
                          >
                            <ForwardRef
                              as="div"
                              onSelect={[Function]}
                            >
                              <div
                                data-reach-menu-item=""
                                data-valuetext="Second"
                                id="option-1--menu--1"
                                onClick={[Function]}
                                onDragStart={[Function]}
                                onFocus={[Function]}
                                onMouseDown={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                onMouseMove={[Function]}
                                onMouseUp={[Function]}
                                role="menuitem"
                                tabIndex={-1}
                              >
                                Second
                              </div>
                            </ForwardRef>
                          </MenuItem>
                        </MenuItem>
                        <MenuItem
                          onSelect={[Function]}
                        >
                          <MenuItem
                            onSelect={[Function]}
                          >
                            <ForwardRef
                              as="div"
                              onSelect={[Function]}
                            >
                              <div
                                data-reach-menu-item=""
                                data-valuetext="Third"
                                id="option-2--menu--1"
                                onClick={[Function]}
                                onDragStart={[Function]}
                                onFocus={[Function]}
                                onMouseDown={[Function]}
                                onMouseEnter={[Function]}
                                onMouseLeave={[Function]}
                                onMouseMove={[Function]}
                                onMouseUp={[Function]}
                                role="menuitem"
                                tabIndex={-1}
                              >
                                Third
                              </div>
                            </ForwardRef>
                          </MenuItem>
                        </MenuItem>
                      </div>
                    </MenuItems>
                  </div>
                </PopoverImpl>
              </Portal>
            </Portal>
          </Popover>
        </MenuPopover>
      </DescendantProvider>
    </DropdownProvider>
  </Menu>
</Menu>
`;
